src/include/kdb.h.in

```diff
@@ -104,16 +110,19 @@ typedef struct _KeySet	KeySet;
  *
  **************************************/
 
-KDB * kdbOpen(Key *errorKey);
+KDB * kdbOpen(const KeySet * contract, Key *parentKey);
+
 int kdbClose(KDB *handle, Key *errorKey);
 
 int kdbGet(KDB *handle, KeySet *returned,
 	Key *parentKey);
 int kdbSet(KDB *handle, KeySet *returned,
 	Key *parentKey);
 
-int kdbEnsure (KDB * handle, KeySet * contract, Key * parentKey);
-
+int elektraGOptsContract (KeySet * contract, int argc, const char * const * argv, const char * const * envp, const Key * parentKey,
```

Comment:
See https://github.com/ElektraInitiative/libelektra/pull/3651#issuecomment-780220801

I tried a separate header at first, but the C++ binding didn't really like that. We can always move the declaration before the 1.0 release. (I think we want to move the header files anyway?)

Reply:
We are already in the process of cleaning up for 1.0 **now**. So we should definitely not introduce something which we know should change before 1.0.

Why did it work for I/O? Why not simply doing it the same way for Opts?

Reply:
I mean including fixing the linkage: having an internal libelektra-opts looks to me like architectural smell (who would be the user of this lib except the plugin). More useful would be to have libelektra-opts which provides an API for using command-line options, which is more than the contract, e.g. it would include a way to get the help message.

Reply:
The I/O and notification contracts are not wrapped by C++ and not exposed in the SWIG bindings (AFAICT using the notification stuff from anything other than C/C++ would be quite difficult). In C++ you would just call the C function directly.

Also, `elektraIoContract` and `elektraNotificationsContract` are implemented separate libraries, so having them in another header is almost a given. In those cases separate libraries also make sense, since you need other functions to interact with the I/O and notification stuff anyway. Just mounting the plugin and letting it do it's thing, like you do with `gopts`, won't really work here.

`libelektra-opts` does provide the full API for using command-line options, including getting the help message. The `gopts` plugin invokes that API with `argc`, `argv` and `envp` (whose values depend on how `gopts` is configured).

Since `libelektra-opts` provides the generated help message via the `parentKey`, which isn't accessible when using `gopts`, we store the help message in a special `proc:/` key when using `gopts`. I don't think a function just for for doing `keyString (ksLookupByName (ks, ..., 0)` would make a lot of sense.

Reply:
I'll try to create a `kdbgopts.h`, but I can't guarantee it will work.

Reply:
I don't mean to provide a function to extract the already formatted message but to remove the `usage` and `prefix` options and have an API call to format the message. Returning by key was a hack because no API was available. Now if we have a reason to have an API (and the contract is a reason), we should rethink the whole approach how to use the options. (And only provide one way, not both gopts and elektraGetOpts.)

Reply:
I don't think just providing `gopts` and making `elektraGetOpts` internal is a good idea. You could use `elektraGetOpts` without mounting a spec. You could even call it entirely without calling `kdbGet`. I know that's not the intended use-case, but we need to maintain the API of `elektraGetOpts` anyway, so why not expose it for potential users?

Also, if your application directly links against `libelektra-opts` you can manually call `elektraGetOptsHelpMessage` using the key `proc:/elektra/gopts/help`. The example `examples/gopts.c` demonstrates that.

However, to make calling that function possible from the various bindings we'd need to create various wrappers (C++, Java, Rust). Using a config for `gopts` and letting the plugin create the help message is just simpler.

Reply:
> I know that's not the intended use-case, but we need to maintain the API of elektraGetOpts anyway, so why not expose it for potential users?

No, we do not necessarily need to maintain this as public API and actually we should not provide any APIs without a clear use-case. Not providing alternative APIs has many advantages, the main advantage: users cannot choose a wrong/unsuitable API by accident. Elektra has too many hidden features. This mostly leads to frustration.

> The example examples/gopts.c demonstrates that.

Thank you for pointing this out.

> However, to make calling that function possible from the various bindings we'd need to create various wrappers (C++, Java, Rust).

Yes, you are right, let us forget about this function. So the remaining question is how we can have the contract functions without polluting the core too much and still have some easy way to construct the contracts in all bindings.

Reply:
> No, we do not necessarily need to maintain this as public API

Strictly speaking we do not need to maintain these signatures:

```c
int elektraGetOpts (KeySet * ks, int argc, const char ** argv, const char ** envp, Key * parentKey);
char * elektraGetOptsHelpMessage (Key * errorKey, const char * usage, const char * prefix);
```

But we do need to maintain the command-line parser, so why not expose this.

A compromise would be to declare that

"We do not provide any compatibility guarantees between releases for the symbols declared in `kdbopts.h`. We strongly recommend to implement command-line parsing via `gopts` and a `kdbOpen()` contract whenever possible."

This way it's possible to use `elektraGetOpts` directly (e.g. for a simple tool that is written once and doesn't need long-term compatibility), but most applications should use the fully supported and maintained `gopts` contract solution.


> Elektra has too many hidden features. This mostly leads to frustration.

Well that is mostly a documentation issue.

Reply:
It is not a documentation issue, it is a problem of having lots of features that do not interact in ways someone would expect. elektraGetOpts is such a function which normally never does what someone would expect, if it would, the whole gopts would not be necessary. That is why I urge again and again that we should get rid of it and provide one way to get proc: keys, a way that works for applications. (And this way would be with gopts (probably rename to proc) via kdbOpen contract.) At the moment many places point to elektraGetOpts as the way to do it, which is very problematic. Declarations that we do not care about maintaining it properly does not help Elektra at all.

Reply:
> elektraGetOpts is such a function which normally never does what someone would expect

What do you expect? Based on what do you expect that?

The function has pretty much the same name as POSIX `getopt` (plus our prefix), so I would assume it has something to do with command-line options. But names in C can be confusing, so let's check the doxygen brief:

https://github.com/ElektraInitiative/libelektra/blob/d3fdba3c36642de3f883dc182e9b9dda377d7a67/src/libs/opts/opts.c#L101-L102

Okay, so it is indeed an extended version of `getopt`. Nothing unexpected here, it just parses the command-line options that's it.

Nothing anywhere gives the impression that this function would validate things like `type`. It also doesn't take a `KDB *` just a plain `KeySet *`, so I wouldn't expect that it interacts with the KDB or plugins in any way.

> At the moment many places point to elektraGetOpts as the way to do it, which is very problematic.

So it is a documentation issue.

Reply:
E.g. I would expect that it works together with spec keys from `kdbGet`.

Reply:
You can call `kdbGet` to get a spec with `opts` metadata and then pass the KeySet to `elektraGetOpts`. That works fine. You won't get any validation for `type` metadata etc. on the Keys generated by `elektraGetOpts`, but why would you expect that to happen? You called the function AFTER `kdbGet` and didn't provide a `KDB` handle, so how would `elektraGetOpts` no anything about what plugins are mounted and need to be excuted?

Reply:
This argumentation is logical from our point of view but if someone simply searches for "how can I elektrify command-line options" (s)he will not think of implications that `kdbGet` needs to be at the "correct" place. And actually no place is correct except the one where gopts executes it. So I argue we should not support any other place, as it only leads to troubles (either spec: keys are not considered or validation does not work).

I see your point that elektraGetOpts is nice because it resembles POSIX getopt. But letting the user call it leads to many subtle differences how applications behave: in ways POSIX getopt could not be misused.

Reply:
Again I see this as a documentation issue. If the documentation clearly states that "elektrifying command-line options" is done via `gopts` and the `kdbOpen` contract. Then I don't see a problem. That is as long as our documentation is structured well enough that this is the first solution you find.

I would still expose `elektraGetOpts` as public API and either just have the doxygen documentation or maybe explain it in an "Advanced Usage" section.

Reply:
Having bloated APIs is not only a documentation issue, it is a much bigger problem. In particular, with docu you can never avoid misguidance because you have no control over where people start reading. E.g. think of someone who first looks at examples/opts.c

What exactly do you expect from providing elektraGetOpts? The only use case I see is for applications that want only command-line options/envs parsing but actually do not want to participate in the vision of Elektra (abstraction of the whole configuration, i.e., having no control over where the configuration comes from). Why do you see such use cases important to support?

If your actual problem is that you want (command-line) specifications elegantly included in the application, we should provide general means to include specifications (like we have support in the high-level API with code generation). I agree that the hacks in examples/gopts.c are not beautiful and there is need for self-contained applications (especially in the migration phase).

Reply:
> elektraCalloc..

I do not see how users can misuse this function. Inside of Elektra they might provide a benefit if we decide to switch to another allocator. It's not that I am a big fan of these functions, they were also not introduced by me. But now starting to not use them would only introduce yet another inconsistency. Removing all of them would be a lot of effort for nearly no gain.

> Sure, but I don't see this as bloated API. If anything the opts stuff is one of the least bloated APIs that Elektra has.
> Yes, these functions are very little effort to maintain, but so is elektraGetOpts.

The bloat and maintenance effort is mostly the increase of the number of libraries we have. I would very much appreciate if we can decrease the libs by one.

> > think of someone who first looks at examples/opts.c
>That just indicates to me that our documentation has absolutely failed. Why would you ever start with a random C code example?

I do not understand what you mean. Why would it be weird if someone clones our repo and does "cd examples &&  vi opts.c" if (s)he is interested in doing something with command-line options?

> So I decided to write my own config tool.

Ok, I see that it could be useful for config tools that want command-line parsing (and do not want config files).

> Especially, if we say elektraGetOpts is public but not stable API.

You are repeating the "we do not care so much about our API stability" mentality. I am against that mentality. Public libraries without stable APIs might introduce a lot of harm. We could install these libs to some private folder.

> The fact that the "vision of Elektra" basically requires a complicated install process of mounting stuff, reminds me much more of Windows than of UNIX.

The vision itself does not require a "complicated install process", you can say the current implementation does if you consider calling `kdb mount` as complicated. With a well-maintained multifile resolver it would be possible to mount by installing a file. But we are going offtopic here.

Reply:
> Why would it be weird if someone clones our repo and does "cd examples && vi opts.c" if (s)he is interested in doing something with command-line options?

Because just cloning a repo without knowing anything about it and starting to copy code from random example you found, is a **very verry** bad idea. Especially in C. At the very least I'd assume people to read the comments in the example code. And like I said, these should point out that this is an advanced example and you should take a look at the documentation first to understand when to use this code.

> You are repeating the "we do not care so much about our API stability" mentality. I am against that mentality. Public libraries without stable APIs might introduce a lot of harm.

Of course I'd prefer stable public APIs as well. I suggested this approach, because it seem to me that you were saying maintaining `elektraGetOpts` as stable API is not worth it.

I would also say that in this case it is not so much of a problem. If someone needs a stable API for processing command-line options they can use `gopts`.

I would basically explain it like this:

- The core behaviour of the command-line options processor is stable. When you write a spec with `opts` type metadata, we guarantee that it will be processed the same way now and in future versions.
- To access the processor you should call it via `gopts`. The plugin provides a stable API. Calling the plugin with a certain configuration will result in the same `proc:/` keys now and in future versions.
- In between there is `elektraGetOpts`. It is externally callable, but we do not give any guarantees that the function signature will remain the same in future versions. You can call this function, if you have a good reason to do so. But your code may not be directly compatible with future versions of Elektra.

In a sense `elektraGetOpts` would be like many of the functions in `kdbprivate.h`. Technically available to external callers (i.e. public API), but not strictly part of our stable public API. To indicate this more clearly, we could use `libelektraprivate_1.0` instead of `libelektra_0.8` in `src/libs/opts/symbols.map`.

Reply:
Ok, I agree, then let us:

> Because just cloning a repo without knowing anything about it and starting to copy code from random example you found, is a very verry bad idea. Especially in C. At the very least I'd assume people to read the comments in the example code. And like I said, these should point out that this is an advanced example and you should take a look at the documentation first to understand when to use this code.

1. write appropriate comments in front of these files that users should prefer the gopts way (with links to the tutorial, that only describes the gopts way) and the way in the example is for "command-line-options-only purposes" without any KDB.
2. Modify or remove examples/optsSnippets.c accordingly, as it shows what not to do.

> In a sense elektraGetOpts would be like many of the functions in kdbprivate.h. Technically available to external callers (i.e. public API), but not strictly part of our stable public API. To indicate this more clearly, we could use libelektraprivate_1.0 instead of libelektra_0.8 in src/libs/opts/symbols.map.

3. clearly mark the header file as private. We should make an elektra/private folder where you can put elektra/private/opts.h
4. :+1: > To indicate this more clearly, we could use libelektraprivate_1.0 instead of libelektra_0.8 in src/libs/opts/symbols.map.